plugins/arm_a64/a64_utils.ml

```diff
@@ -0,0 +1,113 @@
+open Bap_core_theory
+open A64_defs
+open A64_exceptions
+
+module Env = A64_env.Env
+
+module Utils(Core : Theory.Core) = struct
+  open Core
+
+  let skip : Theory.ctrl Theory.eff = perform Theory.Effect.Sort.bot
+  let pass : Theory.data Theory.eff = perform Theory.Effect.Sort.bot
+
+  let is_64bit_register (r: registers_t) = 
+    match r with
+    | `X0 | `X1 | `X2 | `X3 | `X4 | `X5 | `X6 | `X7 | `X8 | `X9
+    | `X10 | `X11 | `X12 | `X13 | `X14 | `X15 | `X16 | `X17 | `X18 | `X19
+    | `X20 | `X21 | `X22 | `X23 | `X24 | `X25 | `X26 | `X27 | `X28 | `X29 
+    | `X30 |`XZR | `SP | `PC | `ELR | `SPSR -> true
+    | _ -> false
+
+
+  let is_64bit_register_op (operand: operand_t) = 
+    match operand with
+    | `Reg r -> is_64bit_register r
+    | `Imm i -> raise (Failure "todo")
+
+  let const_bv64_of (n: int) = int Env.bv64_sort (QWord.int n)
+
+  let bv_sort_of_size (size: int) = Theory.Bitv.define size 
+
+  let const_0bv_of_size (size: int) = zero (bv_sort_of_size size) 
+
+  let load_register_raw (r: registers_t) =
+    let open Env in
+    match r with
+    | `X0 | `W0 -> x0
+    | `X1 | `W1 -> x1
+    | `X2 | `W2 -> x2
+    | `X3 | `W3 -> x3
+    | `X4 | `W4 -> x4
+    | `X5 | `W5 -> x5
+    | `X6 | `W6 -> x6
+    | `X7 | `W7 -> x7
+    | `X8 | `W8 -> x8
+    | `X9 | `W9 -> x9
+    | `X10 | `W10 -> x10
+    | `X11 | `W11 -> x11
+    | `X12 | `W12 -> x12
+    | `X13 | `W13 -> x13
+    | `X14 | `W14 -> x14
+    | `X15 | `W15 -> x15
+    | `X16 | `W16 -> x16
+    | `X17 | `W17 -> x17
+    | `X18 | `W18 -> x18
+    | `X19 | `W19 -> x19
+    | `X20 | `W20 -> x20
+    | `X21 | `W21 -> x21
+    | `X22 | `W22 -> x22
+    | `X23 | `W23 -> x23
+    | `X24 | `W24 -> x24
+    | `X25 | `W25 -> x25
+    | `X26 | `W26 -> x26
+    | `X27 | `W27 -> x27
+    | `X28 | `W28 -> x28
+    | `X29 | `W29 -> x29
+    | `X30 | `W30 -> x30
+    | `XZR | `WZR -> xzr
+    | `SP | `WSP -> sp
+    | `PC -> pc
+    | `ELR -> elr
+    | `SPSR -> spsr
+
+  let load_register (r: operand_t) = 
+    let open Env in
+    match r with
+    | `Reg rr -> load_register_raw rr
+    | `Imm _ -> raise Unexpected_Situation
+
+  let load_immediate (r: operand_t) = 
+    let open Env in
+    match r with
+    | `Imm imm -> int Env.bv32_sort imm
+    | `Reg _ -> raise Unexpected_Situation
+
+  let little_load_bv ~bv_sort ~address = 
+    loadw bv_sort b0 (var Env.memory) address
+
+  let extend_raw ~assem_reg_sort ~unsign ~x = 
+    let extend = if unsign then unsigned else signed in
+    extend assem_reg_sort x 
+
+  let extend_reg ~r ~extend_type ~is_64_reg ~shift_len = 
```

Comment:
For the reference, here's the [reference implementation][1]: 

```
/ ExtendReg()
// ===========
// Perform a register extension and shift

bits(N) ExtendReg(integer reg, ExtendType exttype, integer shift)
    assert shift >= 0 && shift <= 4;
    bits(N) val = X[reg];
    boolean unsigned;
    integer len;

    case exttype of
        when ExtendType_SXTB unsigned = FALSE; len = 8;
        when ExtendType_SXTH unsigned = FALSE; len = 16;
        when ExtendType_SXTW unsigned = FALSE; len = 32;
        when ExtendType_SXTX unsigned = FALSE; len = 64;
        when ExtendType_UXTB unsigned = TRUE;  len = 8;
        when ExtendType_UXTH unsigned = TRUE;  len = 16;
        when ExtendType_UXTW unsigned = TRUE;  len = 32;
        when ExtendType_UXTX unsigned = TRUE;  len = 64;

    // Note the extended width of the intermediate value and
    // that sign extension occurs from bit <len+shift-1>, not
    // from bit <len-1>. This is equivalent to the instruction
    //   [SU]BFIZ Rtmp, Rreg, #shift, #len
    // It may also be seen as a sign/zero extend followed by a shift:
    //   LSL(Extend(val<len-1:0>, N, unsigned), shift);

    len = Min(len, N - shift);
    return Extend(val<len-1:0> : Zeros(shift), N, unsigned);
```

I like very much that your implementation follows closely the reference so we can match it one-to-one and verify its correctness. 

A few ways to improve it:

1) do not abuse labeled parameter (i.e., `~`) they will just clutter the code in the place of use (and we want it to be as close to the specification pseudocode as possible). Especially in this function, they are not needed, as all arguments are of different types so we can't mess with them at the call site. Also, whenever you find that you want to add labels to your arguments, it is a good indicator that the function is not simple anymore and has too complicated interface. 

2) the `is_64_reg` parameter is superfluous here and may introduce potential anomalies. The reason is that the register, which is passed to the function, already has the sort (and correspondingly size) attached to it. So you should use the information provided by the register. 

3) there is a mismatch between the `extend_type` from the reference and our representation. Why is `LSL` there? We need to rectify this asap.

4) at the epilogue of the function we start to diverge from the reference. You shall use the [extend][3] function here, and I fail to see a reason why it could be used here. 
```

// Extend()
// ========

bits(N) Extend(bits(M) x, integer N, boolean unsigned)
    return if unsigned then ZeroExtend(x, N) else SignExtend(x, N);

// Extend()
// ========

bits(N) Extend(bits(M) x, boolean unsigned)
    return Extend(x, N, unsigned);
```
Don't forget that you can always get the sort for any value or register, so you can go, `_ bitv => _ sort => int` and from `_ var => _ sort => int` and the other way around, `int -> _ sort => _ bitv`, etc. We commonly define the following operator (which we should probably add to the Core Theory, given how useful it is),


Basically, here's I would define this function, 
```ocaml
  let extend_reg reg etype shift =
    let s = Theory.Var.sort reg in
    let n = Theory.Bitv.size s in
    let unsigned,len = match etype with
      | SXTB -> false,8
      | SXTH -> false,16
      | SXTW -> false,32
      | SXTX -> false,64
      | UXTB -> true,8
      | UXTH -> true,16
      | UXTW -> true,32
      | UXTX -> true,64 in
    let len = Int.min len (n - shift) in
    let zeros = zero (Theory.Bitv.define shift) in
    let s' = Theory.Bitv.define (len + shift) in
    let v = append s' (static_extract (len-1) 0 (var reg)) zeros in
    extend v s unsigned
```

I had to define a few helpers, first of all, we have the [extract][4] function which is a little bit more generic than we need for our purposes. The Theory's extract takes runtime values for the hi and lo bits of the range of bytes that are extracted but in our case they are static constants defined by the code itself. So we will define a helper called `static_extract` that will take to ML integers (denoting statically known boundaries) and use the generic extract to extract the range,
```
  let static_extract hi lo x =
    let s = Theory.Bitv.define (hi-lo+1) in
    extract s (byte hi) (byte lo) x
```
The `extract` operation requires us to provide the sort of the output value, which we can statically determine as `hi-lo+1`. As you can see, we can define sorts in runtime, as long as you can compute the size of the bitvector. We also need to wrap ML integers into the target language bitvectors. Since we know that they are going to be small (they denote the register sizes) and since the type of `extract`,
```
    val extract : 'a Bitv.t Value.sort -> 'b bitv -> 'b bitv -> _ bitv -> 'a bitv
```
doesn't require `hi` and `lo` to be of the same sort as the extractee, we will define some helpers (which we will certainly use a lot in other places) to create small bitvectors from integers. We pick 8 bits for small integers and add the following definitions to our theory,
```
  module Byte = Bitvec.M8
  let bytes = Theory.Bitv.define 8

  let byte x = int bytes (Byte.int x)

  let static_extract hi lo x =
    let s = Theory.Bitv.define (hi-lo+1) in
    extract s (byte hi) (byte lo) x
```
which corresponds to 

Here, module `Byte` are ML bitvectors with modulus 8 arithmetic. (Do not confuse `Bitv`, which is the Theory (aka the target language) bitvectors with `Bitvec` wich is ML (aka OCaml, aka the host language) bitvectors aka words, the latter are concrete representation the former is an abstraction).  We also defined `bytes` as the sort for bytes (you may find a better place for it). And the small helper `byte` that takes an ML integer and lifts it to the Theory byte value that represents a byte constant with the given bits.

 One more comment about this final expression of our `extract_reg` function
```
    let s' = Theory.Bitv.define (len + shift) in
    let v = append s' (static_extract (len-1) 0 (var reg)) zeros in
    extend v s unsigned
```
which corresponds to
```
 return Extend(val<len-1:0> : Zeros(shift), N, unsigned);
```

our notation is a little bit heavier: 
- `x<n-1:0>` we represent as `static_extract (n-1) 0 x`, 
- `x:y` we represent as `append s x y`, where `s` is the result sort, which in our case is `len+shift`.

You can see that many generic operators in Core Theory are parameterized by sorts which you have to provide explicitly to ensure that they are statically (from the perspective of the target language).


Side question: is your setup OK or you need any help? Like does merlin work for you? Showing type annotation, indicating no errors (when there are none), jumping to the definition (including your own definitions), etc? Indenter working automatically? It is very important that tools are not failing you so that you can focus on the main topic without spending your focus on things that can be done by automation. So if anything is wrong, just tell what and we will guide you how to set it up.


[1]: https://developer.arm.com/docs/ddi0596/f/shared-pseudocode-functions/aarch64-instrsextendreg-pseudocode
[2]: https://developer.arm.com/docs/ddi0596/f/shared-pseudocode-functions/aarch64-instrsextendreg-pseudocode#ExtendType
[3]: https://developer.arm.com/docs/ddi0596/f/shared-pseudocode-functions/shared-functionscommon-pseudocode#impl-shared.Extend.3
[4]: http://binaryanalysisplatform.github.io/bap/api/master/bap-core-theory/Bap_core_theory/Theory/module-type-Basic/index.html#val-extract

Reply:
Hi @ivg, thanks for your detailed directions! 

For the setup, I use VSCode on Ubuntu and have installed a plugin just called OCaml.
>  Like does merlin work for you? Showing type annotation, indicating no errors (when there are none), jumping to the definition (including your own definitions), etc? 

These things works well. One awkward thing is when I write a function in file A and use it in file B, this function will not be recognized(with read wave underline). And the "red wave underline" will only disappear after `bapbuild`.

> Indenter working automatically?

It can automatically indent, but does not fit the ocp-indent format you need. Now I just have a shell file to run `cd ../../` then `make indent` then `cd ./plugins/arm_a64/`.

Reply:
Hi @ivg, thanks for your detailed directions! 

For the setup, I use VSCode on Ubuntu and have installed a plugin just called OCaml.
>  Like does merlin work for you? Showing type annotation, indicating no errors (when there are none), jumping to the definition (including your own definitions), etc? 

These things works well. One awkward thing is when I write a function in file A and use it in file B, this function will not be recognized(with read wave underline). And the "red wave underline" will only disappear after `bapbuild`.

> Indenter working automatically?

It can automatically indent, but does not fit the ocp-indent format you need. Now I just have a shell file to run `cd ../../` then `make indent` then `cd ./plugins/arm_a64/`.

Reply:
@ivg Another question. We have one function per instruction, but what are the parameters of the function? 

Still use instruction **`C6.2.133 LDR (register)`** as an example.  It seems that using a bit vector with 32 bits as an input is the best fit. Because we have pseudo-code like this:

**Decode for all variants of this encoding**
```
integer scale = UInt(size);
if option<1> == '0' then UNDEFINED; // sub-word index
ExtendType extend_type = DecodeRegExtend(option);
integer shift = if S == '1' then scale else 0;
```

`size`, `option` and `S` are all bits in the 32-bit instruction encoding. We can simply extract them out and use them. But the thing is, we are lifting from assembly code, which is essentially characters. So I am confused now.

By the way, how to deal with the UNDEFINED?

Reply:
> there is a mismatch between the extend_type from the reference and our representation. Why is LSL there? We need to rectify this asap.

In the description of `C6.2.133 LDR (register)` (sorry I do not know how to add a link for this), there is a field "Assembler symbols" says:

> <extend> Is the index extend/shift specifier, defaulting to LSL, and which must be omitted for the LSL option
when <amount> is omitted. encoded in the "option" field. It can have the following values:
UXTW when option = 010
LSL when option = 011
SXTW when option = 110
SXTX when option = 111

It is strange, but it does can have LSL, but at here its behavior is same as UXTX when register size is 64 bits and UXTW when 32 bits. So I write code like this [LSL](https://github.com/BinaryAnalysisPlatform/bap/blob/1d848484fc20f0d07202804a02e9eedb028d6cba/plugins/arm_a64/a64_utils.ml#L96).

Reply:
Also, when I try to implement [CheckSPAlignment](https://developer.arm.com/docs/ddi0596/f/shared-pseudocode-functions/aarch64-functionsmemory-pseudocode#CheckSPAlignment), I need to define a lot of things:
```
  let bv64s = Theory.Bitv.define 64
  let bvs_of_size (size: int) = Theory.Bitv.define size

  type bv1
  type bv2
  type bv4
  type bv8

  let bv_of (s: int) (n: int) = 
    let module M = Bitvec.Make(struct let modulus = Bitvec.modulus s end) in
    int (bvs_of_size s) (M.int n)

  let _EL3 = bv_of 2 3
  let _EL2 = bv_of 2 2
  let _EL1 = bv_of 2 1
  let _EL0 = bv_of 2 0

  let sctlr_el1 = Theory.Var.define bv64s "SCTLR_EL1"
  let sctlr_el2 = Theory.Var.define bv64s "SCTLR_EL2"
  let sctlr_el3 = Theory.Var.define bv64s "SCTLR_EL3"

  type proc_state = {
    _N: bv1 Theory.bitv;
    _Z: bv1 Theory.bitv;
    _C: bv1 Theory.bitv;
    _V: bv1 Theory.bitv;
    _D: bv1 Theory.bitv;
    _A: bv1 Theory.bitv;
    _I: bv1 Theory.bitv;
    _F: bv1 Theory.bitv;
    _SP: bv1 Theory.bitv;
    _EL: bv2 Theory.bitv
    (* todo more *)
  }

  let _PSTATE: proc_state = {
    _N = zero (bvs_of_size 1); (* maybe some other default value *)
    _Z = zero (bvs_of_size 1);
    _C = zero (bvs_of_size 1);
    _V = zero (bvs_of_size 1);
    _D = zero (bvs_of_size 1);
    _A = zero (bvs_of_size 1);
    _I = zero (bvs_of_size 1);
    _F = zero (bvs_of_size 1);
    _SP = zero (bvs_of_size 1);
    _EL = zero (bvs_of_size 2)
  }

  let sctlr_el1 = Theory.Var.define bv64s "SCTLR_EL1"
  let sctlr_el2 = Theory.Var.define bv64s "SCTLR_EL2"
  let sctlr_el3 = Theory.Var.define bv64s "SCTLR_EL3"

  let sctlr regime = 
    match regime with
    | _EL1 -> sctlr_el1
    | _EL2 -> sctlr_el2
    | _EL3 -> sctlr_el3
    | _ -> raise (Failure "unreachable")

  let have_el = (* TODO *)

  let s1_translation_regime el = 
    if not is_zero (logxor el _EL0) then el
    else if (* TODO *)

  let check_SP_alignment = 
    if is_zero (logxor _PSTATE._EL _EL0) then (* TODO *)
```
1. I try to model the [ProcState](https://developer.arm.com/docs/ddi0596/f/shared-pseudocode-functions/shared-functionssystem-pseudocode#ProcState) just like the pseudo-code, and I found it is awkward to compare two bit vectors. Now I use `is_zero (logxor a b)` as `==`. Is there any elegant way? I think it should have but I cannot find it :( 
2. I found the way to define bit vector types like ` _N: bv1 Theory.bitv;` is awkward. What do think of my defines? These are important because these things and the code style will exist everywhere in the lifter. 
3. Am I on the right way? Because I found this function seems "endless-nested" and a lot of small functions to be implemented.

Reply:
> `size`, `option` and `S` are all bits in the 32-bit instruction encoding. We can simply extract them out and use them. But the thing is, we are lifting from assembly code, which is essentially characters. So I am confused now.

This part is handled by the LLVM-MC library so we are already handled with decoded parts of the instructions, see the output of `llvm-mc` or `bap-mc`.

> By the way, how to deal with the UNDEFINED? Raise a failure?

Certainly not! Any valid sequence of bits is an expected input to the lifter/disassembler. And undefined expression has [unk][1] semantics. To represent an undefined side-effect, you can use [perform][2]

[1]: https://binaryanalysisplatform.github.io/bap/api/master/bap-core-theory/Bap_core_theory/Theory/module-type-Core/index.html#val-unk
[2]: https://binaryanalysisplatform.github.io/bap/api/master/bap-core-theory/Bap_core_theory/Theory/module-type-Core/index.html#val-perform

Reply:
> This part is handled by the LLVM-MC library so we are already handled with decoded parts of the instructions, see the output of llvm-mc or bap-mc.

I know this. What I mean is that we are lifting from assembly code and the assembly code is characters like `LDR <Wt>, [<Xn|SP>, (<Wm>|<Xm>){, <extend> {<amount>}}]`. So, we still need a step to transfer these characters to those variant:
```
integer scale = UInt(size);
if option<1> == '0' then UNDEFINED; // sub-word index
ExtendType extend_type = DecodeRegExtend(option);
integer shift = if S == '1' then scale else 0;
```
And then pass them to a function like this to implement [LDR (register)](https://developer.arm.com/docs/ddi0596/f/base-instructions-alphabetic-order/ldr-register-load-register-register):
```
let ldr_register scale option extend_type shift regsize datasize = 
  bits(64) offset = ExtendReg(m, extend_type, shift); if HaveMTEExt() then
  SetTagCheckedInstruction(TRUE); bits(64) address;
  bits(datasize) data;
  if n == 31 then 
    CheckSPAlignment(); 
    address = SP[];
  else
    address = X[n];
  address = address + offset;
  data = Mem[address, datasize DIV 8, AccType_NORMAL];
  X[t] = ZeroExtend(data, regsize);
```
I just want to say this step is not close to the specification and not that obvious to be implemented(although, yes, we can interpret those variants from the characters' instruction). I want to check if this step is needed.

Reply:
>   I try to model the [ProcState](https://developer.arm.com/docs/ddi0596/f/shared-pseudocode-functions/shared-functionssystem-pseudocode#ProcState) just like the pseudo-code, and I found it is awkward to compare two bit vectors. Now I use `is_zero (logxor a b)` as `==`. Is there any elegant way? I think it should have but I cannot find it :(

We have the full set of ordering operations, including equality and comparison. 

> I found the way to define bit vector types like _N: bv1 Theory.bitv; is awkward. What do think of my defines? These are important because these things and the code style will exist everywhere in the lifter.

Leading underscores are having special meaning for OCaml compiler, it means that the value is not used, so it won't trigger the warning if it is really unused. So please do not start any of your identifiers with an undescore, unless you really not going to use them. 

Also, don't pack CPU state into an OCaml structure, you're confusing here the meta language (ML) and the target language (that we model). The CPU state is a record in the target language so we shall represent it is a record in the target language. We don't have any special provisions for structures or bit-fields in Core Theory but we commonly model CPU state using a single word and access/modify its bits using extract and bit-wise logic operations. 

> let bv_of (s: int) (n: int) = 
    let module M = Bitvec.Make(struct let modulus = Bitvec.modulus s end) in
    int (bvs_of_size s) (M.int n)

In case when you don't know the modulus beforehand (which is a rare case) you could use the generic interface that works for any modulus, e.g.,
```
let small ~width x = Bitvec.(int x mod modulus width)
```

Though it is rare that you have unknown and fully parameterized witdh, in most cases, e.g., you can use the existing modulus, e.g., to create a one-bit bitvector of an integer you can use `Bitvec.(int x mod m1)`. Also, there are `Bitvec.one` and `Bitvec.zero` for such well-known constants. 

> Also, when I try to implement CheckSPAlignment, I need to define a lot of things:

Please don't implement it, we don't need that level of detail. You can either totally ignore this function or model it as,
```
let aligment_check = Theory.Effect.Sort.ctrl "Aarch64.SPAlignmentCheck"
```
and then just
```
perform alignment_check
```

This will preserve enough information for us so that we can implement this check in runtime, but won't clutter the semantics with service details. 



[1]: https://binaryanalysisplatform.github.io/bap/api/master/bap-core-theory/Bap_core_theory/Theory/module-type-Core/index.html#val-eq
[2]: https://binaryanalysisplatform.github.io/bap/api/master/bitvec/Bitvec/index.html#val-int

Reply:
> > This part is handled by the LLVM-MC library so we are already handled with decoded parts of the instructions, see the output of llvm-mc or bap-mc.
> 
> I know this. What I mean is that we are lifting from assembly code and the assembly code is characters like `LDR <Wt>, [<Xn|SP>, (<Wm>|<Xm>){, <extend> {<amount>}}]`. So, we still need a step to transfer these characters to those variant:
> 
> ```
> integer scale = UInt(size);
> if option<1> == '0' then UNDEFINED; // sub-word index
> ExtendType extend_type = DecodeRegExtend(option);
> integer shift = if S == '1' then scale else 0;
> ```
> 
> And then pass them to a function like this to implement [LDR (register)](https://developer.arm.com/docs/ddi0596/f/base-instructions-alphabetic-order/ldr-register-load-register-register):
> 
> ```
> let ldr_register scale option extend_type shift regsize datasize = 
>   bits(64) offset = ExtendReg(m, extend_type, shift); if HaveMTEExt() then
>   SetTagCheckedInstruction(TRUE); bits(64) address;
>   bits(datasize) data;
>   if n == 31 then 
>     CheckSPAlignment(); 
>     address = SP[];
>   else
>     address = X[n];
>   address = address + offset;
>   data = Mem[address, datasize DIV 8, AccType_NORMAL];
>   X[t] = ZeroExtend(data, regsize);
> ```
> 
> I just want to say this step is not close to the specification and not that obvious to be implemented(although, yes, we can interpret those variants from the characters' instruction). I want to check if this step is needed.

Just to prevent any confusion, let's say it clear that we are not lifting from the assembly (which is indeed a string of characters) but from the LLVM Machine Code (MC), represented as [Disasm_expert.Basic.Insn][1], which is an algebraic data type and is an n-tuple of opcode and a number of operands that are a [sum type][2] of registers, bitvectors, and floating-point constants. So no characters are involved here. Subfields of an instruction encoding, when are relevant are usually provided as immediate operands and we commonly translate them to sum types that more or less correspond to the types in the original specification. We can translate directly in the semantic definition function if such a translation is needed only once (in that case probably even defining a type is not necessary) or we can write a helper function when the variant is used in more than one instruction.  

I  think that we should stick to `tid -> op array -> insn knowledge` as the type for all our semantic definition functions. The lifter will maintain a hashtable from strings, that represent opcodes, to semantic functions of type `tid -> op array -> insn knowledge`. There will be a driver function, that will extract from the knowledge base the MC instruction and select a proper semantic function based on opcode. It will also handle exceptions (e.g., from array indexing operations) and map them to undefined semantics (or even mark them as data in the knowledge base). 

As a side note, it is possible that LLVM decoder misses some of the fields or decodes them incorrectly, in that case, we can always retrieve the memory representation and access any bits of it. 


[1]: https://binaryanalysisplatform.github.io/bap/api/master/bap/Bap/Std/Disasm_expert/Basic/Insn/index.html
[2]: https://binaryanalysisplatform.github.io/bap/api/master/bap/Bap/Std/Op/index.html#type-t